0017-sampler-interface.md

```diff
@@ -0,0 +1,282 @@
+
+```python
+class BitArray(ShapedMixin):
+    """Stores bit-array outcomes."""
+
+    def __init__(self, array: NDArray, num_bits: int):
+        """
+        Args:
+            array: The data, where the last axis is over packed bits whose byte order is 
+                big-endian, the second last axis is over shots, and the preceding axes 
+                correspond to the shape of the experiment.
+            num_bits: How many bit are in each outcome; the size of the bit-array register.
+        """
+        self._array = np.array(array, copy=False, dtype=np.uint8)
+        if self._array.ndim < 2:
+            raise ValueError("The data arary needs to be at least 2-D")
+        self._num_bits = num_bits
+        if self._array.shape[-1] * 8 < self._num_Bits:
+            raise ValueError("The last axis of the data array is not big enough.")
+        # second last dimension is shots, last dimension is packed bits
+        self._shape = self._array.shape[:-2]
+
+    def __repr__(self):
+        desc = f"<num_samples={self.num_shots}, num_bits={self.num_bits}, shape={self.shape}>"
+        return f"BitArray({desc})"
+
+    @property
+    def array(self) -> NDArray[np.uint8]:
+        """The raw array of data."""
+        return self._array
+
+    @property
+    def num_bits(self) -> int:
+        """The number bits in each outcome."""
+        return self._num_bits
+
+    @property
+    def num_samples(self) -> int:
+        """The number of samples."""
+        return self._array.shape[-2]
+
+    def get_counts(self, loc: Tuple[int,...] | None = None, marginal_idxs: Tuple[int,...] | None = None) -> Counts:
+            def get_counts(self, idx: Tuple[int, ...] | None = None) -> Counts:
+        """Join all bit-array output registers together into a new counts structure.
+        
+        Args:
+            idx: Which task coordinate to fetch.
+
+        Raises:
+            ValueError: If no ``idx`` is given and this task result is more than 0-D.
+        """
+```
+
+### Measurement Levels
+
+`backend.run` has the notion of measurement levels which specify whether measurement instructions cause traces (not generallly implemented), complex phases, or discriminated bits to be returned.
+Eventually, measurement levels should be specified inside of a circuit itself by measuring into complex valued registers, as described in the OpenQASM 3 specification.
+This would effectively enable a mix-and-match between different measurement levels, and also allow a circuit to determine its own output types.
+
+However, before this becomes available, in order to acheive feature parity with `backend.run`, we propose implementing this as a flag (similar to `backend.run`) whose value changes the type of all bit-array outputs from `BitArray`s to NumPy arrays with `dtype=np.complex128`.
+
+```python
+class MeasLevel(enum.Enum):
+    # TRACES = 0 unsupported
+    PHASES = 1
+    BITS = 2
+```
```

Comment:
Since this is an IBM extension, does it need to be in the `Sampler` interface definition?  There's no need for the flag for feature parity given the Aer implementation, for example.

Reply:
It seemed pertinent given the claim that we'd like to phase out `backend.run`, and also because we're pointing out an explicit exception to the policy "we obey program outputs".

Reply:
In some ways, it is an OpenPulse extension rather than just an IBM extension. This whole section of the proposal comes from trying to build a bridge toward a future state where Qiskit could represent measurements that return `complex[float[64]]` or `waveform[...]` types. When we have proper support for those types and associated measurement instructions, then this flag could disappear entirely.

Point taken that this _could_ be relegated to IBM's implementation of the Sampler.

Reply:
It's not clear to me that the IBM-only extension of `meas_level` should get such special treatment in the base interface (especially since it wasn't part of the base interface before), and it's really quite worrying that it fundamentally changes the output type of the run method.  Having it as a required option means that all Samplers now need to support it, including those based on backends that didn't previously support it.

Is there a way to go abstractly from a raw-shots return value to the higher measurement-level return value?  If so, then can `meas_level` be a client-side postprocessing step instead of needing quite serious special-casing (including making return types inconsistent and breaking one of the core policies of Sampler) throughout the interface?  If not, then this feels like something that should be outside the remit of a Sampler.  An interface is something we need to live with for a long time if we want to drive adoption of it both from users and from other hardware vendors.  Including this, if it can't be abstracted, would pass on a lot of pain to hardware vendors, including meaning that we'd need to make even in-house projects like Aer and Terra's internal primitives newly support it.

Reply:
> When we have proper support for those types and associated measurement instructions, then this flag could disappear entirely.

If it's in the base interface, it can't disappear easily from the interface; if `SamplerV2` defines that `run` takes it, then we can't remove it from future `SamplerV2`s without breaking user code that uses higher measurement levels, and syncing with other implementers of Sampler.

---

(My "It's not clear to me that the IBM-only ..." comment was written before I saw your comment, Blake - the core of my argument is the same though if it's an OpenPulse or IBM extension, because in practice it's only `IBMProvider` backends that have the field right now.)

Reply:
I think there's been a misunderstanding, likely arising from how it's written. This section exists to allay fears that IBM-only implementations won't have access to meas_level=1 at all. It is not making a comment on the base implementation, and the argument `meas_level=...` was in reference to how we currently allow run() to overwrite options temporarily. I appreciate this latter point is confusing because the three RFCs have said slightly different things on this front, but that is just a side-effect of writing four simultaneously.

The idea that an option value is capable of changing the output type should be treated as a separate issue. The rationale is that we are explicitly accepting a concession that users who want to use that flag are giving up portability during this intermediate period where we can't have the things we want.

Reply:
Given that it's explicitly giving up portability, it might be cleaner to remove all mention of it from the RFC (which by my understanding is of the abstract interface, not specifically for "IBM's Sampler as a replacement of IBM's `backend.run` - disagreement on this might be where the misunderstanding is coming from), and potentially even have the IBM Sampler have a completely separate `run_with_meas_level` function, so that we don't have to give up the strong typing of the base interface.

We'll already be giving up portability by saying that "`meas_level`" can affect the output types, so it feels safer to me to have that be a (temporary) IBM extension that uses a completely different method, which really drives home to users that this is not portable and is a stop-gap helper transition.